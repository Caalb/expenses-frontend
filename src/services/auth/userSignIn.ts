import type { AxiosResponse } from 'axios'
import api from '../api'

export interface UserSignInResponse {
  token: string
}

export const userSignIn = async (
  email: string,
  password: string,
): Promise<AxiosResponse<UserSignInResponse>> => {
  const response = await api.post<UserSignInResponse>('/auth/signin', { email, password })

  return response
}
