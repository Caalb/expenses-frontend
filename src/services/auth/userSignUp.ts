import type { AxiosResponse } from 'axios'
import api from '../api'

export interface UserSignUpResponse {
  token: string
}

export const userSignUp = async (
  name: string,
  email: string,
  password: string,
): Promise<AxiosResponse<UserSignUpResponse>> => {
  const response = await api.post<UserSignUpResponse>('/auth/signup', { name, email, password })

  return response
}
