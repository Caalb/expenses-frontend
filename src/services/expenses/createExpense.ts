import type { Expense } from '@/types/expense'
import api from '@/services/api'

export const createExpense = async (expense: Omit<Expense, 'id'>) => {
  const response = await api.post('/expenses', expense)

  return response
}
