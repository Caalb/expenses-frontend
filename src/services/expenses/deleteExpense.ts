import api from '@/services/api'

export const deleteExpense = async (id: number) => {
  const response = await api.delete(`/expenses/${id}`)

  return response
}
