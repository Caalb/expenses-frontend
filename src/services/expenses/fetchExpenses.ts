import api from '@/services/api'
import type { Expense } from '@/types/expense'

export const fetchExpenses = async (): Promise<Expense[]> => {
  const response = await api.get<Expense[]>('/expenses')

  return response.data
}
