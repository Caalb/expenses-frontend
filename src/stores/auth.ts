import { defineStore } from 'pinia'
import { userSignIn } from '../services/auth/userSignIn'
import type { AxiosResponse } from 'axios'
import { userSignUp } from '@/services/auth/userSignUp'

export const useAuthStore = defineStore('auth', () => {
  const handleSignIn = async (
    email: string,
    password: string,
  ): Promise<AxiosResponse | unknown> => {
    try {
      const response = await userSignIn(email, password)
      const { token } = response.data

      localStorage.setItem('token', token)
      return response
    } catch (error: unknown) {
      return error
    }
  }

  const handleSignUp = async (data: Record<string, string>): Promise<AxiosResponse | unknown> => {
    try {
      const response = await userSignUp(data.name, data.email, data.password)
      const { token } = response.data

      localStorage.setItem('token', token)
      return response
    } catch (error: unknown) {
      return error
    }
  }

  return {
    handleSignIn,
    handleSignUp,
  }
})
