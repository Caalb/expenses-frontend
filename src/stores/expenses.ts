import { createExpense as createExpenseApi } from '@/services/expenses/createExpense'
import { fetchExpenses as fetchExpensesApi } from '@/services/expenses/fetchExpenses'
import type { Expense } from '@/types/expense'
import { defineStore } from 'pinia'
import { ref } from 'vue'

export const useExpensesStore = defineStore('expenses', () => {
  const expenses = ref<Expense[]>([])

  const fetchExpenses = async () => {
    try {
      const response = await fetchExpensesApi()

      expenses.value = response

      return response
    } catch (error) {
      return error
    }
  }

  const createExpense = async (expense: Omit<Expense, 'id'>) => {
    try {
      const response = await createExpenseApi(expense)
      expenses.value = [response.data, ...expenses.value]

      return response
    } catch (error) {
      return error
    }
  }

  return {
    expenses,
    fetchExpenses,
    createExpense,
  }
})
