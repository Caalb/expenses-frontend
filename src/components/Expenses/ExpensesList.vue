<template>
  <div class="q-pa-md">
    <div class="row items-center justify-between q-mb-md">
      <div class="text-h6">Minhas Despesas</div>
      <QBtn
        color="primary"
        icon="add"
        label="Nova Despesa"
        @click="openCreateModal"
      />
    </div>

    <ExpenseTable
      :expenses
      :loading
      @edit="openEditModal"
      @delete="openDeleteModal"
    />

    <QDialog v-model="showFormModal" @hide="resetForm">
      <QCard style="min-width: 400px">
        <QCardSection class="row items-center q-pb-none">
          <div class="text-h6">{{ isEditing ? 'Editar' : 'Nova' }} Despesa</div>
          <QSpace />

          <QBtn icon="close" flat round dense @click="showFormModal = false" />
        </QCardSection>

        <QCardSection>
          <ExpenseForm
            :initial-values="formData"
            :is-editing="isEditing"
            :submitting="submitting"
            @submit="onSubmit"
            @cancel="showFormModal = false"
          />
        </QCardSection>
      </QCard>
    </QDialog>

    <DeleteExpenseDialog
      v-model="showDeleteModal"
      :expense="selectedExpense"
      :loading="deleting"
      @confirm="handleDeleteExpense"
    />
  </div>
</template>

<script setup lang="ts">
import { ref, onMounted, type Ref} from 'vue'
import { useQuasar } from 'quasar'
import { QDialog, QCard, QCardSection, QBtn, QSpace } from 'quasar'
import ExpenseTable from '@/components/Expenses/ExpensesTable.vue'
import ExpenseForm from '@/components/Expenses/ExpenseForm.vue'
import DeleteExpenseDialog from '@/components/Expenses/DeleteExpenseDialog.vue'
import type { Expense, ExpenseFormData } from '@/types/expense'
import { useExpensesStore } from '@/stores/expenses'
import { errorHandler } from '@/helpers/errorHandler'
import type { AxiosError, AxiosResponse } from 'axios'

const $q = useQuasar()
const { expenses, fetchExpenses, createExpense, updateExpense, deleteExpense} = useExpensesStore()

const loading = ref(false)
const submitting = ref(false)
const deleting = ref(false)
const showFormModal = ref(false)
const showDeleteModal = ref(false)
const isEditing = ref(false)
const selectedExpense = ref<Expense | null>(null)
const formData = ref<ExpenseFormData>({
  description: '',
  date: '',
  amount: 0
})


const openCreateModal = () => {
  isEditing.value = false
  formData.value = {
    description: '',
    date: '',
    amount: 0
  }

  showFormModal.value = true
}

const openEditModal = (expense: Expense) => {
  isEditing.value = true
  selectedExpense.value = expense
  formData.value = {
    description: expense.description,
    date: expense.date,
    amount: expense.amount
  }

  showFormModal.value = true
}

const openDeleteModal = (expense: Expense) => {
  selectedExpense.value = expense
  showDeleteModal.value = true
}

const resetForm = () => {
  isEditing.value = false
  selectedExpense.value = null
  formData.value = {
    description: '',
    date: '',
    amount: 0
  }
}

type ExpenseOperation = (values?: ExpenseFormData) => Promise<AxiosResponse>

const handleExpenseOperation = async ({
  operation,
  successMessage,
  errorMessage,
  values,
  modalRef,
}: {
  operation: ExpenseOperation
  successMessage: string
  errorMessage: string
  values?: ExpenseFormData
  modalRef: Ref<boolean>
}) => {
  const response = await operation(values) as AxiosResponse | AxiosError

  if (response && ~~((response.status ?? 0) / 100) === 2) {
    $q.notify({
      color: 'positive',
      message: successMessage,
      icon: 'check',
    })

    modalRef.value = false
    resetForm()
    return
  }

  errorHandler(response, errorMessage, $q)
}

const handleUpdateExpense = (values: ExpenseFormData) =>
  handleExpenseOperation({
    operation: () => updateExpense(selectedExpense.value?.id as number, values) as Promise<AxiosResponse>,
    successMessage: 'Despesa atualizada com sucesso!',
    errorMessage: 'Falha ao atualizar despesa',
    values,
    modalRef: showFormModal
  })

const handleDeleteExpense = () =>
  handleExpenseOperation({
    operation: () => deleteExpense(selectedExpense.value?.id as number) as Promise<AxiosResponse>,
    successMessage: 'Despesa excluÃ­da com sucesso!',
    errorMessage: 'Falha ao excluir despesa',
    modalRef: showDeleteModal
  })

const handleCreateExpense = (values: ExpenseFormData) =>
  handleExpenseOperation({
    operation: () => createExpense(values) as Promise<AxiosResponse>,
    successMessage: 'Despesa criada com sucesso!',
    errorMessage: 'Falha ao criar despesa',
    values,
    modalRef: showFormModal
  })


const onSubmit = async (values: ExpenseFormData) => {
    submitting.value = true

    if(isEditing.value) {
      return handleUpdateExpense(values)
    }

    handleCreateExpense(values)

    submitting.value = false;
}

onMounted(async () => {
  loading.value = true
  await fetchExpenses()
  loading.value = false
})
</script>
