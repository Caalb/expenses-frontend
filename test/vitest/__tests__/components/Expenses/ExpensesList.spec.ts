import ExpensesList from '@/components/Expenses/ExpensesList.vue'
import { createTestingPinia } from '@pinia/testing'
import { installQuasarPlugin } from '@quasar/quasar-app-extension-testing-unit-vitest'
import { mount, VueWrapper } from '@vue/test-utils'
import { beforeEach, describe, expect, it, vi } from 'vitest'
import { useRouter } from 'vue-router'
import { useAuthStore } from '@/stores/auth'
import { useExpensesStore } from '@/stores/expenses'
import router from '@/router'

installQuasarPlugin()

vi.mock('vue-router')

const createWrapper = () =>
  mount(ExpensesList, {
    global: {
      plugins: [
        createTestingPinia({
          createSpy: vi.fn,
        }),
      ],
      mocks: {
        $router: {
          push: vi.fn(),
        },
      },
    },
  }) as VueWrapper

describe('ExpensesList', () => {
  let wrapper: VueWrapper
  let store: ReturnType<typeof useAuthStore>
  let expensesStore: ReturnType<typeof useExpensesStore>

  vi.mocked(useRouter).mockReturnValue({
    ...router,
    push: vi.fn(),
  })

  beforeEach(() => {
    wrapper = createWrapper()
    store = useAuthStore()
    expensesStore = useExpensesStore()
  })

  it('should match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should call fetchExpenses', async () => {
    await wrapper.vm.$nextTick()
    await wrapper.vm.$nextTick()

    expect(expensesStore.fetchExpenses).toHaveBeenCalled()
  })
})
