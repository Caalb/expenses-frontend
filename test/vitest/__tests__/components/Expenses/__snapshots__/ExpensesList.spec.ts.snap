// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExpensesList > should match snapshot 1`] = `
"<div class="q-pa-md">
  <div class="row items-center justify-between q-mb-md">
    <div class="text-h6">Minhas Despesas</div><button class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--rectangle bg-primary text-white q-btn--actionable q-focusable q-hoverable" tabindex="0" type="button"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><i class="q-icon on-left notranslate material-icons" aria-hidden="true" role="img">add</i><span class="block">Nova Despesa</span></span></button>
  </div>
  <div class="q-card q-card--bordered q-card--flat no-shadow" expenses="">
    <div class="q-table__container q-table--horizontal-separator column no-wrap q-table__card q-table--no-wrap">
      <!---->
      <div class="q-table__middle scroll">
        <table class="q-table">
          <thead>
            <tr>
              <th class="text-left">Descrição</th>
              <th class="text-left">Data</th>
              <th class="text-left">Valor</th>
              <th class="text-center" style="width: 100px;">Ações</th>
            </tr>
          </thead>
          <tbody></tbody>
        </table>
      </div>
      <div class="q-table__bottom row items-center q-table__bottom--nodata">
        <div class="full-width row flex-center q-pa-md text-grey-6"> Nenhuma despesa encontrada </div>
      </div>
    </div>
  </div>
  <!---->
  <!---->
</div>"
`;
